### PR TITLE
feat: replace targets with tools

### PR DESCRIPTION
**Pull request by viktorvoltaire**

### CODE REVIEW

File: .mage/magefile.go [hunk 1]
```diff
@@ -41,14 +29,38 @@ func init() {
...
 	)
 	mg.SerialDeps(
-		mggo.GoModTidy,
+		GoModtidy,
 		mggitverifynodiff.GitVerifyNoDiff,
 	)
 }
+
+func FormatYaml() error {
+	return mgyamlfmt.FormatYamlCommand()
```
Comment (odsod) [inline .mage/magefile.go:45]: IMO since this is not a command it's better named `FormatYaml`

Comment (odsod) [reply .mage/magefile.go:45]: actually let's make that `FormatYAML`

File: .mage/magefile.go [hunk 2]
```diff
@@ -41,14 +29,38 @@ func init() {
...
+}
+
+func GoModtidy() error {
```
Comment (odsod) [inline .mage/magefile.go:48]: `GoModTidy`? (capital `T`)

File: .mage/magefile.go [hunk 3]
```diff
@@ -41,14 +29,38 @@ func init() {
...
+	return mggo.GoModTidy().Run()
+}
+
+func Goreview(ctx context.Context) error {
+	return mggoreview.Command(ctx, "-c", "1", "./...").Run()
```
Comment (odsod) [inline .mage/magefile.go:53]: really liking this pattern

File: example/.mage/magefile.go [hunk 4]
```diff
@@ -38,8 +38,8 @@ func init() {
 
 func All() {
 	mg.Deps(
-		mg.F(mgconvco.ConvcoCheck, "origin/master..HEAD"),
-		mggolangcilint.GolangciLint,
+		mg.F(mgconvco.CheckCommand, "origin/master..HEAD"),
```
Comment (odsod) [inline example/.mage/magefile.go:41]: this should be moved to a separate function?

File: mgtool/exec.go [hunk 5]
```diff
@@ -37,3 +39,14 @@ func run(cmd, dir string, stdout, stderr io.Writer, args ...string) error {
...
 }
+
+func Command(path string, args ...string) *exec.Cmd {
+	cmd := exec.Command(path)
+	cmd.Args = append(cmd.Args, args...)
+	cmd.Dir = mgpath.FromGitRoot(".")
+	cmd.Env = os.Environ()
+	cmd.Stderr = os.Stderr
+	cmd.Stdout = os.Stdout
+	cmd.Stdin = os.Stdin
```
Comment (odsod) [inline mgtool/exec.go:50]: IMO this might not be what we want

Comment (odsod) [reply mgtool/exec.go:50]: imagine multiple commands running in parallel, all trying to read from the parent process stdin

File: targets/mgbuf/targets.go [hunk 6]
```diff
@@ -3,49 +3,32 @@ package mgbuf
 import (
 	"context"
 	"fmt"
+	"go.einride.tech/mage-tools/mglog"
```
Comment (odsod) [inline targets/mgbuf/targets.go:6]: formatting

File: targets/mggitverifynodiff/targets.go [hunk 7]
```diff
@@ -1,17 +1,25 @@
...
+	"go.einride.tech/mage-tools/mgtool"
 )
 
 func GitVerifyNoDiff() error {
 	mglog.Logger("git-verify-no-diff").Info("verifying that git has no diff...")
-	diff, _ := sh.Output("git", "status", "--porcelain")
-	if diff != "" {
-		_ = sh.RunV("git", "diff", "--patch")
+	cmd := mgtool.Command("git", "status", "--porcelain")
+	b := &bytes.Buffer{}
```
Comment (odsod) [inline targets/mggitverifynodiff/targets.go:14]: could be simplified to `var b bytes.Buffer` and `cmd.Stdout = &b`

File: targets/mgwire/targets.go [hunk 8]
```diff
@@ -1,11 +1,10 @@
...
-	"github.com/magefile/mage/sh"
-	"go.einride.tech/mage-tools/mglog"
+	"go.einride.tech/mage-tools/mgtool"
+	"os/exec"
 )
 
-func WireGenerate(path string) error {
-	mglog.Logger("wire-generate").Info("generating initializers...")
-	return sh.RunV("go", "run", "-mod=mod", "github.com/google/wire/cmd/wire", "gen", path)
+func WireGenerate(path string) *exec.Cmd {
```
Comment (odsod) [inline targets/mgwire/targets.go:8]: IMO this might be better formulated as

```go
func Command(args ...string) *exec.Cmd {
    args = append(args, "go", "run", "-mod=mod", "github.com/google/wire/cmd/wire")
    return mgtool.Command(args)
}
```

File: .mage/magefile.go [hunk 9]
```diff
@@ -41,14 +28,38 @@ func init() {
...
+		FormatYaml,
 	)
 	mg.SerialDeps(
-		mggo.GoModTidy,
+		GoModtidy,
 		mggitverifynodiff.GitVerifyNoDiff,
 	)
 }
+
+func FormatYaml() error {
```
Comment (Edholm) [inline .mage/magefile.go:43]: Are these functions added for removing the magic `// mage:import` ?

in any case, an added bonus is that you don't need to specify an argument when running only e.g. convco-check

File: .mage/magefile.go (hunk 3 above)
Comment (Edholm) [reply .mage/magefile.go:52]: Me too :100:

File: .mage/magefile.go [hunk 10]
```diff
@@ -41,14 +28,38 @@ func init() {
...
+func Goreview(ctx context.Context) error {
+	return mggoreview.Command(ctx, "-c", "1", "./...").Run()
+}
+
+func GolangciLint(ctx context.Context) error {
+	return mggolangcilint.LintCommand(ctx).Run()
+}
+
+func FormatMarkdown(ctx context.Context) error {
+	return mgmarkdownfmt.Command(ctx, "-w", ".").Run()
```
Comment (Edholm) [inline .mage/magefile.go:60]: Does this command have long flag names? :thinking: If so, I think we should switch. Makes it easier to read and get a feeling for what the command does.

File: mgtool/exec.go [hunk 11]
```diff
@@ -37,3 +39,14 @@ func run(cmd, dir string, stdout, stderr io.Writer, args ...string) error {
 
 	return c.Run()
 }
+
+func Command(path string, args ...string) *exec.Cmd {
```
Comment (Edholm) [inline mgtool/exec.go:43]: I would like to see a comment/docs for this function since it is going to be so ubiquitous.

File: targets/mgapilinter/targets.go [hunk 12]
```diff
@@ -17,23 +17,17 @@ import (
 const version = "1.29.3"
 
 // nolint: gochecknoglobals
-var executable string
+var commandPath string
```
Comment (Edholm) [inline targets/mgapilinter/targets.go:20]: If we really want, we should be able to remove this global variable.

I'm thinking of the following:
* Specify constants for binary name, version, etc
* Add a function that builds the path based of those constants and any variables, like path to tools dir

Could be a future improvement.

File: targets/mggo/targets.go [hunk 13]
```diff
@@ -2,23 +2,20 @@ package mggo
...
 )
 
-func GoTest() error {
+func GoTest() *exec.Cmd {
 	const toolName = "go"
-	mglog.Logger("go-test").Info("running Go unit tests..")
 	coverFile := filepath.Join(mgpath.Tools(), toolName, "coverage", "go-test.txt")
 	if err := os.MkdirAll(filepath.Dir(coverFile), 0o755); err != nil {
-		return err
+		panic(err)
```
Comment (Edholm) [inline targets/mggo/targets.go:16]: image

:joy:

**Review by Edholm — COMMENTED** [review]
Very nice! I like the new structure

File: .mage/magefile.go (hunk 9 above)
Comment (odsod) [reply .mage/magefile.go:43]: With the new structure for running commands centered around `exec.Cmd`, most of the public functions in the tools packages are no longer valid mage targets, so to keep the mental model simple we're thinking that we should instead avoiding having ready-made mage-targets altogether and instead write those in the `magefile.go`.

I think this will lead to a simpler mental model, since there will be a 1-1 mapping between Make targets and Go functions in `magefile.go`.

In short, yes, no more `//mage:import`!

File: .mage/magefile.go (hunk 10 above)
Comment (odsod) [reply .mage/magefile.go:60]: I agree, but unfortunately not: https://github.com/shurcooL/markdownfmt/blob/master/main.go#L23

File: .mage/magefile.go (hunk 9 above)
Comment (Edholm) [reply .mage/magefile.go:43]: Nice, I like it!

File: targets/mgapilinter/targets.go (hunk 12 above)
Comment (odsod) [reply targets/mgapilinter/targets.go:20]: Really like this!

File: .mage/magefile.go (hunk 9 above)
Comment (odsod) [reply .mage/magefile.go:43]: @viktorvoltaire Minor thing, but we should rename the `targets` folder to `tools` - since it now contains tools and not targets. :smile_cat:

File: targets/mgapilinter/targets.go (hunk 12 above)
Comment (viktorvoltaire) [reply targets/mgapilinter/targets.go:20]: Actually, given that we added symlinks to all tools we download, the path is now deterministic. Should be a fairly simple change.

File: example/.mage/magefile.go (hunk 4 above)
Comment (viktorvoltaire) [reply example/.mage/magefile.go:41]: Yeah I haven’t touched this one yet

File: targets/mgapilinter/targets.go (hunk 12 above)
Comment (viktorvoltaire) [reply targets/mgapilinter/targets.go:20]: However, i will look at that in a separate PR

File: mgtool/file.go [hunk 14]
```diff
@@ -73,6 +79,12 @@ func FromLocal(ctx context.Context, filepath string, opts ...Opt) error {
 	if s.skipFile != "" {
 		// Check if binary already exist
 		if _, err := os.Stat(s.skipFile); err == nil {
+			if s.symlink != "" {
+				_, err := CreateSymlink(s.symlink)
+				if err != nil {
```
Comment (odsod) [inline mgtool/file.go:84]: join the if statement with the assignment above

**Review by odsod — APPROVED** [review]
Just when we thought it couldn't get any better :smile:

**Review by odsod — APPROVED** [review]
:shipit: :rocket: :rainbow:
